### DISCUSSION TITLE
How integrate with goconvey?

### DISCUSSION THREAD

**Discussion by aleus** [gavv/httpexpect#143] [Q&A] [answered]
I want to use the best features of both best test-frameworks: [goconvey](https://github.com/smartystreets/goconvey) and [httpexpect](https://github.com/gavv/httpexpect). But both frameworks use its own print format. Does anybody have example of integrate httpexpect in goconvey.

**Comment by gavv** [comment]
Hi,

httpexpect always uses these interfaces to print anything:

```
// Printer is used to print requests and responses.
// CompactPrinter, DebugPrinter, and CurlPrinter implement this interface.
type Printer interface {
	// Request is called before request is sent.
	Request(*http.Request)

	// Response is called after response is received.
	Response(*http.Response, time.Duration)
}

// Logger is used as output backend for Printer.
// testing.TB implements this interface.
type Logger interface {
	// Logf writes message to log.
	Logf(fmt string, args ...interface{})
}

// Reporter is used to report failures.
// testing.TB, AssertReporter, and RequireReporter implement this interface.
type Reporter interface {
	// Errorf reports failure.
	// Allowed to return normally or terminate test using t.FailNow().
	Errorf(message string, args ...interface{})
}
```

You can provide your own implementation using corresponding `Config` fields. I didn't use goconvey, but I guess you could just implement these interfaces using convey.Print*.

It's enough to implement just Printer and Reporter, however if you want to reuse default httpexpect printers formatting (CompactPrinter, DebugPrinter, or CurlPrinter), you can implement Logger instead of Printer and pass your logger to one of existing printers.

**Comment by aleus** [comment]
Thaks. I'll try and I'll say about results.

**Comment by ybbus** [comment]
@aleus I am very interested in the results and the setup. Since I also use Convey. Thanks

**Comment by tyranron** [comment, marked as answer]
At the moment I'm doing the following trick

1) Reporter that causes GoConvey assertion fail:
```go
type ConveyReporter struct {
	convey.C
}

// Implements httpexpect.Reporter interface.
func (c ConveyReporter) Errorf(message string, args ...interface{}) {
	c.C.So(fmt.Sprintf(message, args...), assertionFails)
}

func assertionFails(actual interface{}, _ ...interface{}) string {
	return actual.(string)
}
```

2) Printer that causes GoConvey assertion success:
```go
type ConveyPrinter struct {
	convey.C
}

// Request implements Printer.Request.
func (ConveyPrinter) Request(*http.Request) {
	// Does nothing.
}

// Response implements Printer.Response.
func (p ConveyPrinter) Response(*http.Response, time.Duration) {
	p.C.So(true, convey.ShouldBeTrue)
}
```

3) Using it:
```go
func NewExpect(c convey.C) *Expect {
	return WithConfig(Config{
		BaseURL:  myUrl,
		Reporter: ConveyReporter{c},
		Printers: []Printer{ConveyPrinter{c}},
	})
}
 
// ...
func TestGetUsers(t *testing.T) {
	Convey("GET /users", t, func(c C) {
		NewExpect(c).GET("/users").
			Expect().
			Status(http.StatusBadRequest).
			JSON().Schema(ErrorSchema).
			Object().ValueEqual("code", ErrRequiredAccessToken)
	})
})
```


As the result I have one success GoConvey assertion for every request that has response and if some httpexpect assertion fails it generates GoConvey failure assertion.

@gavv is it possible to dig somehow into httpexpect assertion engine? I want to bind every httpexpect assertion to appropriate GoConvey assertion. Even successful ones, not only failed.

**Comment by gavv** [comment]
@tyranron currently there is no single entry point for every assertion. To accomplish that, one need to patch all the code and register every issued assertion in somewhere. I guess we could just extend the Reporter interface with a Success() method and call this method from every check in httpexpect. It would be quite a large patch though the changes would be mechanical.

**Comment by gavv** [comment]
> @gavv is it possible to dig somehow into httpexpect assertion engine? I want to bind every httpexpect assertion to appropriate GoConvey assertion. Even successful ones, not only failed.

It's now possible. See here https://github.com/gavv/httpexpect#customize-assertion-handling
